autogtp/main.cpp

```diff
@@ -221,10 +224,23 @@ int main(int argc, char *argv[])
 
     auto success = true;
     auto games_played = 0;
+    auto retries = 0;
 
     do {
         QString netname;
         success &= fetch_best_network_hash(cerr, netname);
+        if (!success) {
+            if (retries++ >= MAX_RETRIES) {
+                cerr << "Maximum number of retries exceeded. Giving up." << endl;
+                return EXIT_FAILURE;
+            }
+            cerr << "Failed to get the network. Retrying in " <<
+                RETRY_DELAY << " s." << endl;
+            QThread::sleep(RETRY_DELAY);
```

Comment:
With 30s between retries and 1K clients, that's 33 qps - probably OK for serving a static error page. But then when the server recovers all 1K clients will try to download the new best network during the same 30s and then after the first game, submit the new game probably during the ~same few-minute period. That might be a bit annoying.

The best option would be to do exponential backoff instead. 30s would be an OK starting point, but maybe do something like:

```
retry_dealy = std::min(RETRY_DELAY * std::pow(1.5, retry), 3600)
```

That would start from 30s, 45s, 67s ... up to 3600s on the 13th attempt and keep on retrying every hour from that point on. Still recovers in a few minutes from a temporary blip, backs off nicely to allow for the service to recover and tries for as long as it is needed to recover from outages.

For bonus points, also include a randomized component (maybe from 50% of the computed retry delay to 100% of the retry delay) to spread out the stampeding herd. But maybe that's going too far for this change? :)

Reply:
I added the exponential backoff. I don't think the randomization is worth the trouble right now.